hearthstone/simulator/core/card_pool.py

```diff
@@ -2624,6 +2624,27 @@ def handle_event_powers(self, event: CardEvent, context: Union[BuyPhaseContext,
                         context.owner.play_blood_gem(card)
 
 
+class Shudderling(MonsterCard):
+    not_in_pool = True
+    tier = 1
+    monster_type = MONSTER_TYPES.NEUTRAL
+    base_attack = 1
+    base_health = 1
+
+    def base_battlecry(self, targets: List[MonsterCard], context: BuyPhaseContext):
+        battlecry_minion_queue = [card for card in context.owner.minions_played if card.battlecry and type(card) != type(self)]
```

Comment:
What about minions which were played and then sold?  This may be dangerous since the list of `minions_played` might point to things no longer on the board.

Reply:
ah, hmmm. We want any battlecries that modify `self` to now modify this minion. But I also think we need to trigger a golden battlecry if the original minion was golden. How do we do this?

There's a similar issue with the `FishOfNZoth` card, that triggers all previous deathrattles. The deathrattles should be golden if the dead card was golden.

Just realized I forgot the golden bonus for shudderling, which I will add now.

Reply:
You can strip off the "self" target of the battlecry function, and provide it with a different self argument.  There is an example here: https://github.com/JDBumgardner/stone_ground_hearth_battles/blob/master/hearthstone/simulator/core/cards.py#L90 (see how it is called here with the "self" argument? https://github.com/JDBumgardner/stone_ground_hearth_battles/blob/master/hearthstone/simulator/core/combat.py#L229)

But I don't think that answers the original question about cards that are played and then sold?

Reply:
Seems like that might solve the Fish of NZoth issue. Currently the code for the fish is:
```
def handle_event_powers(self, event: 'CardEvent', context: Union['BuyPhaseContext', 'CombatPhaseContext']):
    if event.event is EVENTS.DIES and not event.card == self and event.card.deathrattles and event.card in context.friendly_war_party.board:
        for deathrattle in event.card.deathrattles:
            for _ in range(2 if self.golden else 1):
                self.deathrattles.append(deathrattle)
```
If I'm understanding this correctly, would we do 
```
                self.deathrattles.append(event.card.deathrattle.__func__)
```
and this would mean when Fish of Nzoth calls this deathrattle, the `self` argument is `event.card`?

Reply:
> But I don't think that answers the original question about cards that are played and then sold?

I don't think so. Consider Annihilan Battlemaster, with battlecry
```
def base_battlecry(self, targets: List[MonsterCard], context: BuyPhaseContext):
    multiplier = 2 if self.golden else 1
    damage_taken = context.owner.hero.starting_health() - context.owner.health
    self.health += multiplier * damage_taken
```
If we play a golden battlemaster and then a normal shudderling, `self.golden` should return the `AnnihilanBattlemaster.golden`, while `self.health` should return `Shudderling.health`. So it doesn't seem like the above solution would work here.

Reply:
Wait, I think I'm wrong. Because in `combat.py` we do
```
for deathrattle in minion.deathrattles:
    deathrattle(minion, context)
```
so we'd still be calling the deathrattle with Fish of NZoth as `self`.

(as an aside, why are we calling the function this way instead of doing `minion.deathrattle(context)`?)

Reply:
We do it this way because we need to copy deathrattle around between cards, so we need to separate the logic of the deathrattle from the card it is associated with.

I think the easiest workaround would be to add an additional optional argument to force a battlecry to be golden or not.

Do you know how golden cards are handled with shudderling though? Are the two non-golden battlecries also copied?  I can imagine a problem where the card that gets golden transformed  does a regular battlecry and then a golden battlecry, but shudderling will only copy one of them.  It might make sense to directly track the battlecries rather than just the played minions.  Since the battlecries are just functions, we don't need to worry about them being changed after they are called.

Reply:
I now remember a solution we employed with Replicating Menace's deathrattle:
```
def __init__(self):
      def base_deathrattle(card, context: 'CombatPhaseContext'):
          summon_index = context.friendly_war_party.get_index(card)
          for i in range(3 * context.summon_minion_multiplier()):
              microbot = Microbot()
              if self.golden:
                  microbot.golden_transformation([])
              context.friendly_war_party.summon_in_combat(microbot, context, summon_index + i + 1)

      self.base_deathrattle = types.MethodType(base_deathrattle, self)
      super().__init__()
```
Could we do something similar with Annihilan Battlemaster and other similar minions (I think there are only a couple)?

Reply:
Also, was the solution I had above correct?

Reply:
I'm not sure I understand about the relationship between replicating menace and Annihilan Battlemaster.  If we really do need to check the golden status of the minion, while targeting the "self" of the shuddlerling, then I think we need to write the battlecry functions in such a way that they effectively take two arguments, one for each case (even though most of the time they are the same).  Which solution above are you referring to?

Reply:
We could also hack shudderling to pretend it's golden while calling the battlecries, and then switch it back when we are done.

Reply:
> I'm not sure I understand about the relationship between replicating menace and Annihilan Battlemaster. If we really do need to check the golden status of the minion, while targeting the "self" of the shuddlerling, then I think we need to write the battlecry functions in such a way that they effectively take two arguments, one for each case (even though most of the time they are the same). Which solution above are you referring to?

Just the fact that we are defining the deathrattle to take `card`, not `self`, but we are using the value of `self` in the function. I was thinking we could define the battlecry in a similar way.

Reply:
Ah, I think we didn't worry about it, because battlecries seem relatively immutable, while deathrattles can be copied (i.e. due to magnetism).  If we started copying or manipulating battlecries, then I think it would make sense to do the same thing.

Reply:
It seems to me that the shudderling effect _is_ copying battlecries.

Reply:
Does shudderling act as if it were `self` in the battlecry call? I thought the battlecries were being played "in place"?

Reply:
Wait I was confused, yes we should be storing the battlecry function without the `self` parameter bound, and explicitly passing the parameter.

Reply:
Ok, so I tried doing this by replicating what we did for Replicating Menace:
```
class AnnihilanBattlemaster(MonsterCard):
    ...

    def __init__(self):
        def base_battlecry(card, targets: List[MonsterCard], context: BuyPhaseContext):
            multiplier = 2 if self.golden else 1
            damage_taken = context.owner.hero.starting_health() - context.owner.health
            card.health += multiplier * damage_taken

        self.base_battlecry = types.MethodType(base_battlecry, self)
        super().__init__()
```

and then within the battlecry function for Shudderling, we call Annihilan Battlemaster's battlecry and explicitly pass the `self` parameter:

```
class Shudderling(MonsterCard):
    ...

    def base_battlecry(self, targets: List[MonsterCard], context: BuyPhaseContext):
        battlecry_minion_queue = [card for card in context.owner.minions_played if card.battlecry and type(card) != type(self)]
        for next_minion in battlecry_minion_queue:
            valid_targets = [card for card in context.owner.in_play if next_minion.valid_battlecry_target(card)]
            chosen_targets = []
            for _ in range(next_minion.num_battlecry_targets[-1]):
                target = context.randomizer.select_friendly_minion(valid_targets)
                valid_targets.remove(target)
                chosen_targets.append(target)
            for _ in range(2 if self.golden else 1):
HERE ==>        next_minion.battlecry(self, chosen_targets, context)
```
but I'm getting a `TypeError: base_battlecry() takes 3 positional arguments but 4 were given` when we reach this line.

Can you explain a bit more how to do this? What am I doing wrong?